### PR TITLE
Replace bookmark count metric

### PR DESCRIPTION
**Pull request by DJAndries**
Resolves https://github.com/brave/brave-browser/issues/35937

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

0. Start with fresh profile, wait 40 seconds. Ensure metric value is 0.
1. Add some bookmarks, restart browser.
2. Wait 40 seconds, ensure metric value corresponds to the number of bookmarks saved.
3. Repeat steps 1-2 for different amounts of bookmarks.

### CODE REVIEW

**Comment by DJAndries** [comment]
will switch base to `master` once https://github.com/brave/brave-core/pull/21920 is merged

File: components/p3a/histograms_braveizer.cc [hunk 1]
```diff
@@ -18,7 +18,6 @@ namespace {
 // Please keep this list sorted and synced with |DoHistogramBravezation|.
 // clang-format off
 constexpr const char* kBravezationHistograms[] = {
-    "Bookmarks.Count.OnProfileLoad",
```
Comment (iefremov) [inline components/p3a/histograms_braveizer.cc:21]: cannot we just update it to v3?

File: components/misc_metrics/page_metrics.cc [hunk 2]
```diff
@@ -242,4 +260,13 @@ void PageMetrics::OnDomainDiversityResult(
   VLOG(2) << "PageMetrics: domains loaded report, count = " << count;
 }
 
+void PageMetrics::ReportBookmarkCount() {
+  if (!bookmark_model_ || !bookmark_model_->loaded()) {
+    return;
+  }
+  int count = CountBookmarks(bookmark_model_->root_node());
```
Comment (iefremov) [inline components/misc_metrics/page_metrics.cc:267]: if we really don't like reusing the existing histogram, why don't we use`BookmarkCounter` here

File: components/p3a/histograms_braveizer.cc (hunk 1 above)
Comment (DJAndries) [reply components/p3a/histograms_braveizer.cc:21]: the existing metric broke due to a histogram naming change, and we didn't realize it for quite some time. using a `BookmarkCounter` seems like less of a risk.

File: components/misc_metrics/page_metrics.cc (hunk 2 above)
Comment (DJAndries) [reply components/misc_metrics/page_metrics.cc]: thanks, changed

File: components/misc_metrics/page_metrics.cc [hunk 3]
```diff
@@ -27,17 +29,20 @@ namespace {
 constexpr const int kPagesLoadedBuckets[] = {0, 10, 50, 100, 500, 1000};
 constexpr const int kDomainsLoadedBuckets[] = {0, 4, 10, 30, 50, 100};
 constexpr const int kFailedHTTPSUpgradeBuckets[] = {0, 25, 50, 100, 300, 700};
+constexpr const int kBookmarkCountBuckets[] = {0,   5,    20,   100,
+                                               500, 1000, 5000, 10000};
 
-constexpr base::TimeDelta kReportInterval = base::Minutes(30);
-constexpr base::TimeDelta kInitReportDelay = base::Seconds(30);
+constexpr base::TimeDelta kReportInterval = base::Seconds(10);
```
Comment (iefremov) [inline components/misc_metrics/page_metrics.cc:35]: why seconds?

File: components/misc_metrics/page_metrics.h [hunk 4]
```diff
@@ -35,12 +44,15 @@ inline constexpr char kDomainsLoadedHistogramName[] =
     "Brave.Core.DomainsLoaded";
 inline constexpr char kFailedHTTPSUpgradesHistogramName[] =
     "Brave.Core.FailedHTTPSUpgrades";
+inline constexpr char kBookmarkCountHistogramName[] =
+    "Brave.Core.BookmarkCount";
 
 class PageMetrics {
```
Comment (iefremov) [inline components/misc_metrics/page_metrics.h:50]: this class probably deserves a comment

File: components/misc_metrics/page_metrics.cc [hunk 5]
```diff
@@ -129,6 +142,7 @@ void PageMetrics::ReportAllMetrics() {
   ReportDomainsLoaded();
   ReportPagesLoaded();
   ReportFailedHTTPSUpgrades();
+  ReportBookmarkCount();
   periodic_report_timer_.Start(
       FROM_HERE, base::Time::Now() + kReportInterval,
```
Comment (iefremov) [inline components/misc_metrics/page_metrics.cc:147]: counting bookmarks can be expensive (there are users with thousands of them), so let's avoid doing this every 10 seconds

**Comment by iefremov** [comment]
can we add a test?

File: components/misc_metrics/page_metrics.cc (hunk 5 above)
Comment (DJAndries) [reply components/misc_metrics/page_metrics.cc:147]: my mistake. i changed the intervals for dev purposes and forgot to change them back

**Review by iefremov — APPROVED** [review]

**Pull request merged**